### DISCUSSION TITLE
Need help in defining global variable in Cypress Tests

### DISCUSSION THREAD

**Discussion by mohsinht** [cypress-io/cypress#15966] [Questions and Help] [answered]
I want to use a Global variable inside my Cypress test file but its value isn't changing as expected despite adding waits.

```JS
const builder = {
  stepsArr: []
};
describe('my test', () => {
  beforeEach(() => {
    cy.intercept('/graphql', (req) => {
      req.continue((res) => {
        if (res.body.data?.steps) {
          builder.stepsArr = res.body.data.steps.steps;
          console.log({ stepsArr: builder.stepsArr }); // logs correctly!
        }
      });
    }).as('graphqlRequest');
  });
  it.only('should check global var', () => {
    cy.waitFor('steps');
    cy.wrap({ getStepByTitle: pathwayBuilder.getStepByTitle })
      .invoke('getStepByTitle', {
        steps: builder.stepsArr // always empty array!
      })
      .then((stepObj) => {
        cy.log(stepObj);
      });
  });
});
```

**The order of execution is correct but the value of Global variable isn't updating. Its showing empty array when I invoke my function despite retrying for like 100 times. What could be wrong?**

`cy.waitFor('steps');` is from a command in `support/commands.js` file
```JS
Cypress.Commands.add('waitFor', operationName => {
  cy.wait('@graphqlRequest').then(({ request }) => {
    if (request.body.operationName !== operationName) {
      cy.log('Waiting for:', operationName)
      return cy.waitFor(operationName)
    }
    return null
  })
})
```

The function simply:
```JS
exports.pathwayBuilder = {
    getStepByTitle: (title, steps) => {
        console.log("Search", title);
        console.log("Steps", steps); // empty!
   }
}
```

**Comment by mohsinht** [comment]
For future reference => **Solution**: https://stackoverflow.com/questions/67086884/how-can-i-update-a-global-variable-in-cypress-tests/67087543?noredirect=1#comment118585889_67087543

**Comment by bahmutov** [comment, marked as answer]
I am not a big fan of the solution. In general, I would 
- wrap the data and NOT the logic. In your case, you would want to wrap the 
```js
const builder = {
  stepsArr: []
};
```
and NOT the `cy.wrap({ getStepByTitle: pathwayBuilder.getStepByTitle })`

JavaScript passes the arguments by value, thus when you evaluate the
```js
.invoke('getStepByTitle', {
    steps: builder.stepsArr // always empty array!
})
```

which evaluates right away, the `builder.stepsArr` is empty and will never change. You could move this block into `.then` to get around this problem

```js
cy.waitFor('steps');
  .then(() => {
    cy.wrap({ getStepByTitle: pathwayBuilder.getStepByTitle })
      .invoke('getStepByTitle', {
        steps: builder.stepsArr // always empty array!
      })
      .then((stepObj) => {
        cy.log(stepObj);
      });
  })
```

But a much more elegant solution is to simply wait for the property to exist and then invoke your `pathwayBuilder.getStepByTitle`. I added an example of this to the recipe https://glebbahmutov.com/cypress-examples/7.0.1/recipes/wait-for-data.html

In your case, you can remove the custom command `waitFor` and simply have:

```js
const builder = {};
describe('my test', () => {
  beforeEach(() => {
    cy.intercept('/graphql', (req) => {
      if (res.body.data?.steps) {
          builder.stepsArr = res.body.data.steps.steps;
          console.log({ stepsArr: builder.stepsArr }); // logs correctly!
        }
    })
  });
  it('should check global var', () => {
    cy.wrap(builder)
       .should('have.property', 'stepsArr')
       // have.property assertion yields the value!
       .then(steps => {
          return getStepByTitle(builder.stepsArr)
       })
       // add any other assertions
```

Read the section https://glebbahmutov.com/blog/cypress-tips-and-tricks/#wait-for-data
